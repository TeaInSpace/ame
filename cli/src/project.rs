use ame::client::native_client::AmeClient;
use anyhow::Result;

use ame::grpc::TaskCfg;
use console::Emoji;

use serde::{Deserialize, Serialize};
use std::{fs, fs::File};

#[derive(Clone, Default, Deserialize, Serialize)]
pub struct Project {
    #[serde(rename = "name")]
    id: String,

    #[serde(skip_serializing_if = "Option::is_none")]
    tasks: Option<Vec<TaskCfg>>,

    #[serde(skip_serializing_if = "Option::is_none")]
    templates: Option<Vec<TaskCfg>>,
}

impl Project {
    pub fn new(name: &str) -> Self {
        Project {
            id: name.to_string(),
            ..Project::default()
        }
    }

    pub fn init(name: &str) -> Result<()> {
        if File::open("ame.yaml").is_ok() {
            println!(
                "An AME project is already initiated in this directory. {}",
                Emoji("✨", ":-)")
            );

            Ok(())
        } else {
            let project = Project::new(name);
            let f = File::create("ame.yaml")?;
            serde_yaml::to_writer(f, &project).unwrap();

            println!("Project initialised {}", Emoji("🏗", ""));

            Ok(())
        }
    }

    pub fn task_names(&self) -> Vec<String> {
        self.tasks
            .as_ref()
            .map(|tasks| tasks.iter().filter_map(|t| t.name.clone()).collect())
            .unwrap_or(vec![])
    }
    pub fn init_from_working_dir() -> Result<Self> {
        Ok(serde_yaml::from_str(&fs::read_to_string("ame.yaml")?)?)
    }

    pub fn get_task_template(&self, _name: &str) -> Result<TaskCfg> {
        todo!();
        /*
        let valid_task_templates: Vec<&TaskTemplate> =
            task_templates.iter().filter(|t| t.name == name).collect();

        if valid_task_templates.len() > 1 {
            return Err(Error::MissConfiguredProject(
                "found multiple tasks templates with the same name".to_string(),
            ));
        }

        if valid_task_templates.is_empty() {
            return Err(Error::MissingTaskTemplate(name.to_string()));
        }

        Ok(valid_task_templates[0].clone())
        */
    }

    pub async fn run_task(&self, _client: &mut AmeClient, template_name: &str) -> Result<()> {
        let project_file: Project = serde_yaml::from_str(&fs::read_to_string("ame.yaml")?)?;
        let _task_template = project_file.get_task_template(template_name)?;
        Ok(())

        //     // TODO: handle name clashes in the cluster.
        //     let random_task_name = format!(
        //         "{}{}",
        //         template_name,
        //         Alphanumeric.sample_string(&mut rand::thread_rng(), 6)
        //     )
        //     .to_lowercase();

        //     client
        //         .create_task_project_directory(tonic::Request::new(TaskProjectDirectoryStructure::new(
        //             &random_task_name,
        //             &self.id,
        //             vec![],
        //         )))
        //         .await?;

        //     let _chunk_size = 500;

        //     for entry in walkdir::WalkDir::new(".").into_iter().flatten() {
        //         if entry.metadata()?.is_dir() {
        //             continue;
        //         }

        //         let task_name = random_task_name.clone();

        //         let Ok(mut f) = File::open(entry.clone().path()) else {
        //                         continue;
        //                     };

        //         let mut buf: [u8; 100] = [0; 100];

        //         // TODO: How do we test that files are uploaded transferred correctly? in the
        //         // common ame-client library perhaps?
        //         let stre = async_stream::stream! {

        //             yield ProjectFileChunk {
        //                 messages: Some(Messages::Identifier(ProjectFileIdentifier{
        //                     taskid: task_name,
        //                     filepath:entry.clone().path().to_str().unwrap().to_string(),
        //                 }))
        //             };

        //             loop {

        //         let n = match f.read(&mut buf) {
        //             Ok(0) => {
        //                 break;
        //             }

        //             Ok(n) => n,

        //             // TODO: how do we handle errors here?
        //             Err(_) => {
        //                 break;
        //             },
        //         };

        //         yield  ProjectFileChunk {
        //             messages: Some(Messages::Chunk(FileChunk {
        //                  contents: buf.get(0..n).unwrap().to_vec()
        //             })),
        //          };

        //         buf = [0; 100];
        //          }

        //             };

        //         client.upload_project_file(stre).await?;
        //     }

        //     println!("uploaded project!");

        //     client
        //         .create_task(tonic::Request::new(CreateTaskRequest::new(
        //             &random_task_name,
        //             task_template,
        //         )))
        //         .await?;

        //     tokio::time::sleep(Duration::from_secs(1)).await;

        //     let mut log_stream = client
        //         .stream_task_logs(tonic::Request::new(TaskLogRequest::stream_from_beginning(
        //             &random_task_name,
        //             true,
        //         )))
        //         .await?
        //         .into_inner();

        //     while let Some(entry) = log_stream.next().await {
        //         let Ok(line) = String::from_utf8(entry.clone()?.contents) else {
        //             println!("failed to parse log entry: {entry:?}");
        //             return Ok(());
        //         };

        //         if line.contains("s3") || line.contains("argo") {
        //             continue;
        //         }

        //         if line.contains("WARNING:") {
        //             break;
        //         }

        //         print!("{line}");
        //     }

        //     Ok(())
    }
}
