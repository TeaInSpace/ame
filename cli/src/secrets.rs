use std::time::Duration;

use ame_client::client_builder::{build_ame_client, AmeServiceClientCfg};
use ame_client::Empty;
use ame_client::{AmeSecret, AmeSecretId};
use clap::Subcommand;
use colored::Colorize;
use dialoguer::{Input, Password};
use spinners::Spinner;
use tonic::Request;

use crate::{CliConfiguration, Result};

/// Manage secrets in  secret store.
#[derive(Subcommand)]
pub enum SecretCommand {
    /// Create a secret
    ///
    /// The key will be used to identify the secret, when used with other
    ///  resources.
    ///
    /// You will be prompted to input the secret value, how ever the value will not be
    /// displayed for security.
    Create {
        /// Key used to identify the secret.
        key: Option<String>,
    },

    /// Delete a secret
    Delete {
        /// Key used to identify the secret.
        key: Option<String>,
    },

    /// List all secrets
    List,
}

pub async fn exec_secret_command(cfg: CliConfiguration, cmd: &SecretCommand) -> Result<()> {
    let mut client = build_ame_client(AmeServiceClientCfg {
        disable_tls_cert_check: true,
        endpoint: cfg.endpoint.parse().unwrap(),
        id_token: cfg.id_token,
    })
    .await?;

    match cmd {
        SecretCommand::Create { key } => {
            let secret = AmeSecret {
                key: key.clone().unwrap_or_else(|| {
                    Input::new()
                        .with_prompt("Please provide the secret key")
                        .interact()
                        .unwrap()
                }),
                value: Password::new()
                    .with_prompt("Please provide the secret value")
                    .interact()
                    .unwrap(),
            };

            let mut spinner = Spinner::new(
                spinners::Spinners::Dots,
                format!("{} secret", "Storing".cyan().bold()),
            );

            let mut request = Request::new(secret);
            request.set_timeout(Duration::from_secs(5));

            let res = client.create_secret(request).await;

            match res {
                Ok(_) => {
                    spinner.stop_and_persist(" ", format!("{} secret", "Stored".green().bold()));
                    Ok(())
                }
                Err(e) => {
                    spinner.stop_and_persist(
                        " ",
                        format!("{} to store secret", "Failed".red().bold()),
                    );
                    Err(e)
                }
            }
        }
        SecretCommand::Delete { key } => {
            let secret_id = AmeSecretId {
                key: key.clone().unwrap_or_else(|| {
                    Input::new()
                        .with_prompt("Please provide a secret key")
                        .interact()
                        .unwrap()
                }),
            };

            let mut spinner = Spinner::new(
                spinners::Spinners::Dots,
                format!("{} secret", "Deleting".cyan().bold()),
            );

            match client.delete_secret(Request::new(secret_id)).await {
                Ok(_) => {
                    spinner.stop_and_persist(" ", format!("{} secret", "Deleted".green().bold()));
                    Ok(())
                }
                Err(e) => {
                    spinner.stop_and_persist(
                        " ",
                        format!("{} to delete secret", "Failed".red().bold()),
                    );
                    Err(e)
                }
            }
        }
        SecretCommand::List => {
            let mut spinner = Spinner::new(
                spinners::Spinners::Dots,
                format!("{} secrets", "Fetching".cyan().bold()),
            );
            let secrets = client.list_secrets(Request::new(Empty {})).await?;
            spinner.stop_with_message(format!("{} ", "Key".white().bold()));

            for s in secrets.into_inner().secrets {
                println!("{}", s.key);
            }

            Ok(())
        }
    }?;
    Ok(())
}
