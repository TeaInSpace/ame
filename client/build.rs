fn main() {
    // TODO: evaluate state of optionals in protobuf.
    tonic_build::configure()
        .build_server(false)
        .type_attribute(".", "#[derive(serde::Serialize, serde::Deserialize)]")
        .protoc_arg("--experimental_allow_proto3_optional")
        .compile(&["../proto/ame.proto"], &["../proto"])
        .unwrap_or_else(|e| panic!("Failed to compile protos {:?}", e));
}
