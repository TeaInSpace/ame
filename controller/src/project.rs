use crate::TaskSpec;

use k8s_openapi::apimachinery::pkg::apis::meta::v1::OwnerReference;
use kube::CustomResource;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

#[derive(CustomResource, Deserialize, Serialize, Clone, Debug, JsonSchema, Default)]
#[kube(
    kind = "Project",
    group = "ame.teainspace.com",
    version = "v1alpha1",
    namespaced,
    status = "ProjectStatus",
    shortname = "proj"
)]
#[serde(rename_all = "camelCase")]
pub struct ProjectSpec {
    #[serde(rename = "projectid")]
    id: String,

    #[serde(skip_serializing_if = "Option::is_none")]
    tasks: Option<Vec<TaskSpec>>,

    #[serde(skip_serializing_if = "Option::is_none")]
    templates: Option<Vec<TaskSpec>>,
}

#[derive(Deserialize, Serialize, Clone, Debug, JsonSchema, PartialEq, Eq, Default)]
pub struct ProjectStatus {}

impl Project {
    pub fn add_owner_reference(&mut self, owner_reference: OwnerReference) -> &mut Project {
        match &mut self.metadata.owner_references {
            Some(refs) => refs.push(owner_reference),
            None => self.metadata.owner_references = Some(vec![owner_reference]),
        };

        self
    }
}
