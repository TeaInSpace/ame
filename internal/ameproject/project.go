package ameproject

import (
	"archive/tar"
	"bytes"
	"context"
	"errors"
	"fmt"
	"io"
	"path/filepath"

	"google.golang.org/grpc/metadata"
	"teainspace.com/ame/api/v1alpha1"
	"teainspace.com/ame/cmd/ame/filescanner"
	genv1alpha "teainspace.com/ame/generated/clientset/versioned/typed/ame/v1alpha1"
	"teainspace.com/ame/internal/auth"

	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	task_service "teainspace.com/ame/server/grpc"
	"teainspace.com/ame/server/storage"
)

// TODO: centralize the chunksize configuration
const MdKeyProjectName = "project-name"

func ProjectNameFromDir(dir string) string {
	return filepath.Base(dir)
}

type ProjectConfig struct {
	Name        string
	Directory   string
	AuthToken   string
	ProjectFile *ProjectFileCfg
}

type Project struct {
	ProjectConfig
	taskClient task_service.TaskServiceClient
}

func NewProject(cfg ProjectConfig, taskClient task_service.TaskServiceClient) Project {
	return Project{
		ProjectConfig: cfg,
		taskClient:    taskClient,
	}
}

func (p *Project) GetTaskSpec(name string) (*v1alpha1.TaskSpec, error) {
	for n := range p.ProjectFile.Specs {
		if TaskSpecName(name) == n {
			return p.ProjectFile.Specs[n], nil
		}
	}

	return nil, fmt.Errorf("could not find task specification for %s", name)
}

func (p *Project) UploadAndRunSpec(ctx context.Context, name string) (*v1alpha1.Task, error) {
	spec, err := p.GetTaskSpec(name)
	if err != nil {
		return nil, err
	}

	task := v1alpha1.NewTaskFromSpec(spec, name)

	return p.UploadAndRun(ctx, task)
}

func (p *Project) UploadAndRun(ctx context.Context, t *v1alpha1.Task) (*v1alpha1.Task, error) {
	ctx = auth.AuthorarizeCtx(ctx, p.AuthToken)
	err := p.UploadProject(ctx)
	if err != nil {
		return nil, fmt.Errorf("got error while attempting to upload project: %v", err)
	}

	taskInCluster, err := p.UploadTaskForProject(ctx, t)
	if err != nil {
		return nil, err
	}

	return taskInCluster, nil
}

func (p *Project) UploadTaskForProject(ctx context.Context, t *v1alpha1.Task) (*v1alpha1.Task, error) {
	return p.taskClient.CreateTask(ctx,
		&task_service.TaskCreateRequest{
			Task: t,
		})
}

func (p *Project) UploadProject(ctx context.Context) error {
	// There is no need to upload the ame project file, so it is filtered out.
	t, err := filescanner.TarDirectory(p.Directory, []string{AmeProjectFileName})
	if err != nil {
		return fmt.Errorf("got error while attempting to upload project: %v", err)
	}

	ctx = metadata.AppendToOutgoingContext(ctx, MdKeyProjectName, p.Name)

	uploadClient, err := p.taskClient.FileUpload(ctx)
	if err != nil {
		return err
	}

	send := func(data []byte) error {
		return uploadClient.Send(&task_service.Chunk{
			Contents: data,
		})
	}

	err = task_service.ProcessInChunks(t, send, task_service.ChunkSize)
	if err != nil {
		return fmt.Errorf("got error while transferring chunsk: %v", err)
	}

	_, err = uploadClient.CloseAndRecv()
	return err
}

type LogProcessor func(*task_service.LogEntry) error

func (p *Project) ProcessTaskLogs(ctx context.Context, targetTask *v1alpha1.Task, logProcessor LogProcessor) error {
	ctx = auth.AuthorarizeCtx(ctx, p.AuthToken)
	logsClient, err := p.taskClient.GetLogs(ctx, &task_service.GetLogsRequest{
		Task:       targetTask,
		Follow:     true,
		GetAllLogs: true,
	})
	if err != nil {
		return err
	}

	// TODO determine why this loop is necessary
	for {
		logEntry, err := logsClient.Recv()
		if errors.Is(err, io.EOF) {
			break
		}

		if err != nil {
			return err
		}

		err = logProcessor(logEntry)
		if err != nil {
			return err
		}
	}

	return nil
}

func (p *Project) GetTask(ctx context.Context, taskName string, ns string) (*v1alpha1.Task, error) {
	ctx = auth.AuthorarizeCtx(ctx, p.AuthToken)
	return p.taskClient.GetTask(ctx, &task_service.TaskGetRequest{
		Name:      taskName,
		Namespace: ns,
	})
}

func (p *Project) GetArtifacts(ctx context.Context, taskName string) ([]storage.ProjectFile, error) {
	ctx = auth.AuthorarizeCtx(ctx, p.AuthToken)
	return GetArtifacts(ctx, p.taskClient, taskName)
}

func (p *Project) ScheduleTask(ctx context.Context, spec v1alpha1.TaskSpec, schedule string) (*v1alpha1.ReccurringTask, error) {
	ctx = auth.AuthorarizeCtx(ctx, p.AuthToken)
	return p.taskClient.CreateRecurringTask(ctx, &task_service.RecurringTaskCreateRequest{
		Task: v1alpha1.NewRecurringTask(p.Name, spec, schedule),
	})
}

// TODO: GetArtifacts loads all of the artifacts into memory at once, this will not work for large artifacts.

func GetArtifacts(ctx context.Context, taskClient task_service.TaskServiceClient, taskName string) ([]storage.ProjectFile, error) {
	artifactsClient, err := taskClient.GetArtifacts(ctx, &task_service.ArtifactGetRequest{TaskName: taskName})
	if err != nil {
		return nil, err
	}

	data := []byte{}

	for {
		chunk, err := artifactsClient.Recv()
		if errors.Is(err, io.EOF) {
			break
		}

		if err != nil {
			return nil, err
		}

		data = append(data, chunk.Contents...)
	}

	files := []storage.ProjectFile{}
	err = filescanner.ReadFromTar(bytes.NewBuffer(data), func(h *tar.Header, b []byte) error {
		files = append(files, storage.ProjectFile{
			Data: b,
			Path: h.Name,
		})

		return nil
	})

	if err != nil {
		return nil, err
	}

	return files, nil
}

func GetTasksForProject(ctx context.Context, tasks genv1alpha.TaskInterface, projectId string) ([]v1alpha1.Task, error) {
	taskList, err := tasks.List(ctx, metav1.ListOptions{})
	if err != nil {
		return nil, err
	}

	var projectTasks []v1alpha1.Task
	for _, t := range taskList.Items {
		if t.Spec.ProjectId == projectId {
			projectTasks = append(projectTasks, t)
		}
	}

	return projectTasks, nil
}
