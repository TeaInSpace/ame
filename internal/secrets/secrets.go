// Package secrets implements tools for interacting with AME secrets.
//
// The initial implementation is a simple abstraction over Kubernetes Secrets. The intent is to keep the outward interface
// more or less unchanged as the implementation is developed, so the rest of the codebase can start using this package without
// needing major changes as the implementaion is developed.
package secrets

import (
	"context"

	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	clientgocorev1 "k8s.io/client-go/kubernetes/typed/core/v1"

	corev1 "k8s.io/api/core/v1"
	"k8s.io/apimachinery/pkg/api/errors"
)

// secretKey is the key used by all AME secrets stored as K8s secrets.
const secretKey = "secret"

// An AmeSecret represents a secret that can be used by a Task.
// The Name is used to identify the Secret and Value is contents in the form of a string.
type AmeSecret struct {
	Name  string
	Value string
}

// A SecretStore represents the Store of secrets AME uses to write and read to.
// AME curently stores secrets as Kubernetes secrets.
type SecretStore struct {
	// secrets is the interface to the k8s REST API for secrets.
	secrets clientgocorev1.SecretInterface
}

// NewSecretStore creates a new SecreteStore which uses secrets to communicate
// with the Kubernetes API. What cluster and namespace secrets is configured
// to use, is also what the SecretStore will use.
// The New SecretStore is returned.
func NewSecretStore(secrets clientgocorev1.SecretInterface) SecretStore {
	return SecretStore{
		secrets: secrets,
	}
}

// ForceCreate creates the secret and overwrites any existing secret with the same name.
// If an error occurs it is returned immediately aborting the secret creation.
func (s *SecretStore) ForceCreate(ctx context.Context, secret AmeSecret) error {
	err := s.Create(ctx, secret)
	if errors.IsAlreadyExists(err) {
		err := s.Delete(ctx, secret.Name)
		if err != nil {
			return err
		}

		return s.Create(ctx, secret)
	}

	if err != nil {
		return err
	}

	return nil
}

// Create attempts to create the secret. If there already exists a secret with the same name,
// an error will be returned, and the existing secret is not overwritten.
// If an error occurs it is returned.
func (s *SecretStore) Create(ctx context.Context, secret AmeSecret) error {
	k8sSecret := &corev1.Secret{
		ObjectMeta: metav1.ObjectMeta{
			Name: secret.Name,
		},
		StringData: map[string]string{
			secretKey: secret.Value,
		},
	}
	_, err := s.secrets.Create(ctx, k8sSecret.DeepCopy(), metav1.CreateOptions{})
	return err
}

// Delete attempts to delete a secret identified by the name.
// If an error occurs it is returned.
func (s *SecretStore) Delete(ctx context.Context, name string) error {
	return s.secrets.Delete(ctx, name, metav1.DeleteOptions{})
}

// SecretEnvVarSrc returns an EnvVarSource for the secret identified
// by secretName.
func SecretEnvVarSrc(secretName string) *corev1.EnvVarSource {
	return &corev1.EnvVarSource{
		SecretKeyRef: &corev1.SecretKeySelector{
			Key: secretKey,
			LocalObjectReference: corev1.LocalObjectReference{
				Name: secretName,
			},
		},
	}
}
