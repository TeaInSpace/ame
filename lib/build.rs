use std::io;

fn main() -> io::Result<()> {
    tonic_build::configure()
        .build_server(true)
        .build_client(true)
        .type_attribute(
            ".",
            "#[derive(serde::Serialize, serde::Deserialize, schemars::JsonSchema )]",
        )
        .protoc_arg("--experimental_allow_proto3_optional")
        .compile(&["ame.proto"], &["../proto"])
}
