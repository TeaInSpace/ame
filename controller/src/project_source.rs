use crate::project::{Project, ProjectSpec};
use crate::secrets::SecretCtrl;
use crate::{Error, Result};
use ame::grpc::GitProjectSource;
use ame::grpc::ProjectSourceState;
use ame::grpc::ProjectSourceStatus;
use duration_string::DurationString;
use envconfig::Envconfig;
use futures::{future::BoxFuture, FutureExt, StreamExt};
use git2::build::RepoBuilder;
use git2::{Cred, FetchOptions};
use k8s_openapi::api::core::v1::Secret;
use k8s_openapi::apimachinery::pkg::apis::meta::v1::OwnerReference;
use k8s_openapi::chrono::DateTime;

use kube::{
    api::{Api, ListParams, PatchParams, ResourceExt},
    client::Client,
    core::ObjectMeta,
    runtime::controller::{Action, Controller},
    CustomResource, Resource,
};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use std::path::Path;
use std::time::SystemTime;
use std::{fs, sync::Arc, time::Duration};
use tracing::{info, warn};

#[derive(CustomResource, Deserialize, Serialize, Clone, Debug, JsonSchema, Default)]
#[kube(
    kind = "ProjectSource",
    group = "ame.teainspace.com",
    version = "v1alpha1",
    namespaced,
    status = "ProjectSourceStatus",
    shortname = "psrc"
)]
#[serde(rename_all = "camelCase")]
pub struct ProjectSourceSpec {
    #[serde(flatten)]
    pub cfg: ame::grpc::ProjectSourceCfg,
}

#[derive(Envconfig, Clone)]
pub struct ProjectSrcCtrlCfg {
    #[envconfig(from = "NAMESPACE", default = "ame-system")]
    pub namespace: String,
}

impl ProjectSource {
    async fn git_secret(&self, secrets: Api<Secret>) -> Result<Option<String>> {
        let Some(GitProjectSource {  secret: Some(secret_name), .. }) = self.spec.clone().cfg.git else {
        return Ok(None);
    };
        Ok(Some(
            SecretCtrl::from(secrets).get_secret(&secret_name).await?,
        ))
    }

    async fn extract_projects(&self, secrets: Api<Secret>) -> Result<Vec<ProjectSpec>> {
        let Some(GitProjectSource{
        repository,
        username,
        ..
        }) = self.spec.clone().cfg.git else {
            return Err(Error::MissingProjectSrc("git".to_string()))
        };

        let _ = fs::remove_dir_all("/tmp/".to_string() + &self.name_any());

        let git_secret = self.git_secret(secrets).await?;

        let mut opts = FetchOptions::new();

        if git_secret.is_some() {
            let mut callbacks = git2::RemoteCallbacks::new();
            callbacks.credentials(|_user, _, _| -> std::result::Result<Cred, git2::Error> {
                let Some(git_secret) = git_secret.clone() else {
                    return Err(git2::Error::from_str("git secret was empty"))
                };

                let Some(username) = username.clone() else {
                return Err(git2::Error::from_str("missing username"));

                };

                git2::Cred::userpass_plaintext(&username, &git_secret)
            });

            opts.remote_callbacks(callbacks);
        }

        let mut builder = RepoBuilder::new();
        builder.fetch_options(opts);

        // TODO: ensure that cloning never clashes with other directories.
        // TODO: How will we handle large repositories?
        let _repo = builder.clone(
            &repository,
            Path::new(&format!("/tmp/{}", &self.name_any())),
        )?;

        let project = serde_yaml::from_str(&fs::read_to_string(format!(
            "/tmp/{}/ame.yaml",
            self.name_any()
        ))?)?;

        fs::remove_dir_all("/tmp/".to_string() + &self.name_any())?;

        Ok(vec![project])
    }

    fn sync_interval(&self) -> Result<Duration> {
        if let Some(GitProjectSource {
            sync_interval: Some(sync_internal),
            ..
        }) = &self.spec.cfg.git
        {
            Ok(DurationString::try_from(sync_internal.clone())
                .map_err(Error::InvalidDuration)?
                .into())
        } else {
            Ok(Duration::from_secs(5 * 60))
        }
    }

    fn requires_sync(&self) -> Result<bool> {
        if let Some(ProjectSourceStatus {
            last_synced: Some(ref last_synced),
            ..
        }) = self.status
        {
            Ok(std::time::SystemTime::now()
                .duration_since(SystemTime::from(DateTime::parse_from_rfc2822(last_synced)?))?
                > self.sync_interval()?)
        } else {
            Ok(true)
        }
    }
}

struct Context {
    client: Client,
    config: ProjectSrcCtrlCfg,
}

async fn reconcile(src: Arc<ProjectSource>, ctx: Arc<Context>) -> Result<Action> {
    info!("Reconciling {}", src.name_any());

    let _srcs = Api::<ProjectSource>::namespaced(ctx.client.clone(), &ctx.config.namespace);
    let projects = Api::<Project>::namespaced(ctx.client.clone(), &ctx.config.namespace);
    let oref = if let Some(refe) = src.controller_owner_ref(&()) {
        refe
    } else {
        OwnerReference::default()
    };

    let secrets = Api::<Secret>::namespaced(ctx.client.clone(), &ctx.config.namespace);

    let mut patch: ProjectSource = _srcs.get_status(&src.name_any()).await?;
    patch.metadata.managed_fields = None;

    if src.requires_sync()? {
        let project_specs = match src.extract_projects(secrets).await {
            Ok(specs) => specs,
            Err(e) => {
                let status = ProjectSourceStatus {
                    state: ProjectSourceState::Error.into(),
                    reason: Some(e.to_string()),
                    ..patch.status.unwrap_or(ProjectSourceStatus::default())
                };

                patch.status = Some(status);

                _srcs
                    .patch_status(
                        &src.name_any(),
                        &PatchParams::apply("ame-controller"),
                        &kube::api::Patch::Apply(patch),
                    )
                    .await?;
                return Ok(Action::requeue(Duration::from_secs(50)));
            }
        };

        if project_specs.is_empty() {
            return Ok(Action::requeue(Duration::from_secs(50)));
        }

        let mut project = Project {
            metadata: ObjectMeta {
                name: Some(src.name_any()),
                ..ObjectMeta::default()
            },
            spec: project_specs[0].clone(),
            status: None,
        };

        if let Some(GitProjectSource { ref repository, .. }) = src.spec.cfg.git {
            project.add_annotation("gitrepository".to_string(), repository.to_string());
        }

        let project = project.add_owner_reference(oref);

        projects
            .patch(
                &project.name_any(),
                &PatchParams::apply("ame-controller"),
                &kube::api::Patch::Apply(project),
            )
            .await?;
    }

    let last_synced = Some(humantime::format_rfc3339(SystemTime::now()).to_string());
    let mut patch: ProjectSource = _srcs.get_status(&src.name_any()).await?;
    patch.metadata.managed_fields = None;

    if let Some(mut status) = patch.clone().status {
        status.last_synced = last_synced;
        status.reason = Some("project has been synced".to_string());
        status.state = ProjectSourceState::Synchronized.into();
        patch.status = Some(status);
    } else {
        patch.status = Some(ProjectSourceStatus {
            last_synced,
            reason: Some("project has been synced".to_string()),
            state: ProjectSourceState::Synchronized.into(),
            ..ProjectSourceStatus::default()
        })
    }

    _srcs
        .patch_status(
            &src.name_any(),
            &PatchParams::apply("ame-controller"),
            &kube::api::Patch::Apply(patch),
        )
        .await?;

    Ok(Action::requeue(Duration::from_secs(5 * 60)))
}

fn error_policy(_src: Arc<ProjectSource>, error: &Error, _ctx: Arc<Context>) -> Action {
    warn!("failed to reconcile: {:?}", error);
    Action::requeue(Duration::from_secs(5 * 60))
}

pub async fn start_project_source_controller(config: ProjectSrcCtrlCfg) -> BoxFuture<'static, ()> {
    let client = Client::try_default().await.expect("Failed to create a K8S client, is the controller running in an environment with access to cluster credentials?");
    let context = Arc::new(Context {
        client: client.clone(),
        config,
    });

    let project_srcs = Api::<ProjectSource>::namespaced(client.clone(), &context.config.namespace);
    project_srcs
        .list(&ListParams::default())
        .await
        .expect("Is the CRD installed?");

    let projects = Api::<Project>::namespaced(client, &context.config.namespace);

    Controller::new(project_srcs, ListParams::default())
        .owns(projects, ListParams::default())
        .run(reconcile, error_policy, context)
        .filter_map(|x| async move { std::result::Result::ok(x) })
        .for_each(|_| futures::future::ready(()))
        .boxed()
}

#[cfg(test)]
mod test {

    use crate::common::private_repo_gh_pat;

    use super::*;
    use crate::secrets::SecretCtrl;
    use assert_fs::prelude::*;
    use futures::StreamExt;
    use futures::TryStreamExt;
    use k8s_openapi::api::core::v1::Secret;
    use kube::api::DeleteParams;

    use kube::{api::PostParams, core::WatchEvent};
    use serde_json::json;
    use serial_test::serial;

    fn test_project_src() -> Result<ProjectSource> {
        Ok(serde_json::from_value(json!({
            "apiVersion": "ame.teainspace.com/v1alpha1",
            "kind": "ProjectSource",
            "metadata": { "name": "test" },
            "spec": {
                "git": {
                    "repository": "https://github.com/TeaInSpace/ame-test.git",
                },
            }
        }))?)
    }

    fn private_test_project_src() -> Result<ProjectSource> {
        Ok(serde_json::from_value(json!({
            "apiVersion": "ame.teainspace.com/v1alpha1",
            "kind": "ProjectSource",
            "metadata": { "name": "test-private" },
            "spec": {
                "git": {
                    "repository": "https://github.com/TeaInSpace/ame-test-private.git",
                    "secret": "ghsecret",
                    "username": "jmintb",
                },
            }
        }))?)
    }

    #[tokio::test]
    #[serial]
    async fn can_extract_projects_from_public_git_repository() -> Result<()> {
        let test_dir = assert_fs::TempDir::new().unwrap();
        let working_dir = std::env::current_dir()?;

        std::env::set_current_dir(test_dir.path())?;

        let project_src: ProjectSource = serde_json::from_value(json!({
            "apiVersion": "ame.teainspace.com/v1alpha1",
            "kind": "ProjectSource",
            "metadata": { "name": "test" },
            "spec": {
                "git": {
                    "repository": "https://github.com/TeaInSpace/ame-test.git",
                },
            }
        }))?;

        let client = Client::try_default().await?;
        let secrets = Api::<Secret>::default_namespaced(client);
        let projects = project_src.extract_projects(secrets).await?;
        insta::assert_yaml_snapshot!(&projects);

        test_dir
            .child(project_src.name_any())
            .assert(predicates::path::missing());

        std::env::set_current_dir(working_dir)?;
        Ok(())
    }

    #[tokio::test]
    #[serial]
    async fn can_create_project_from_src() -> Result<()> {
        let client = Client::try_default().await?;
        let project_srcs = Api::<ProjectSource>::default_namespaced(client.clone());
        let projects = Api::<Project>::default_namespaced(client.clone());

        let _handle = tokio::spawn(
            start_project_source_controller(ProjectSrcCtrlCfg {
                namespace: "default".to_string(),
            })
            .await,
        );

        let project_src = test_project_src()?;

        let _ = project_srcs
            .delete(&project_src.name_any(), &DeleteParams::default())
            .await;

        project_srcs
            .create(&PostParams::default(), &project_src)
            .await?;

        let mut stream = projects
            .watch(&ListParams::default().timeout(60), "0")
            .await?
            .boxed();
        while let Ok(Some(e)) = stream.try_next().await {
            if let WatchEvent::Added(_project) = e {
                return Ok(());
            }
        }

        panic!("failed to create project ",);
    }

    #[tokio::test]
    #[serial]
    async fn can_create_project_from_private_src() -> Result<()> {
        let client = Client::try_default().await?;
        let project_srcs = Api::<ProjectSource>::default_namespaced(client.clone());
        let projects = Api::<Project>::default_namespaced(client.clone());

        let secret_ctrl = SecretCtrl::new(client, "default");

        let secret_name = "ghsecret";
        secret_ctrl
            .store_secret_if_empty(secret_name, private_repo_gh_pat().unwrap())
            .await
            .unwrap();

        let _handle = tokio::spawn(
            start_project_source_controller(ProjectSrcCtrlCfg {
                namespace: "default".to_string(),
            })
            .await,
        );

        let mut project_src = private_test_project_src()?;

        let _ = project_srcs
            .delete(&project_src.name_any(), &DeleteParams::default())
            .await;

        project_src = project_srcs
            .create(&PostParams::default(), &project_src)
            .await?;

        let mut stream = projects
            .watch(&ListParams::default().timeout(60), "0")
            .await?
            .boxed();
        while let Ok(Some(e)) = stream.try_next().await {
            if let WatchEvent::Added(project) = e {
                if project
                    .metadata
                    .owner_references
                    .unwrap()
                    .iter()
                    .any(|r| r.name == project_src.name_any())
                {
                    return Ok(());
                }
            }
        }

        panic!("failed to create project ",);
    }

    //TODO test that sync interval can be overridden
    //TODO test that sync works
}
