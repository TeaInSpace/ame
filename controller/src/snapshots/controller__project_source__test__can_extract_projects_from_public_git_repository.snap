---
source: controller/src/project_source.rs
expression: "&projects"
---
- projectid: echo
  tasks:
    - runcommand: python echo.py myoutput
      projectid: echo
      env: ~
      image: ~
      secrets: ~
      pipeline: ~
      source:
        gitrepository: ~
        gitreference: ~
        amestoragepath: ~
      resources:
        cpu: "2"
        memory: 4Gi

